addons/mail/tools/discuss.py

```diff
@@ -182,55 +167,194 @@ def get_result(self):
                 res[model_name] = [dict(sorted(record.items())) for record in records.values()]
         return res
 
-    @staticmethod
-    def many(records, mode="REPLACE", /, *, as_thread=False, only_id=False, **kwargs):
-        """Flags records to be added to the store in a Many relation.
+    def _add_records_data(self, records, fields, as_thread=False):
+        """Fetches data from records for the given fields, and follow relations."""
+        if not records:
+            return []
+        records_data = records._read_format(
+            [field for field in fields if not isinstance(field, Store.Relation)],
+            load=False,
+        )
+        if as_thread:
+            records_data = [{**record_data, "model": records._name} for record_data in records_data]
+        self.add("mail.thread" if as_thread else records._name, records_data)
+        for record in records:
+            data = {}
+            for relation_field in [field for field in fields if isinstance(field, Store.Relation)]:
+                if relation_field.field_name == "thread" and "thread" not in record._fields:
+                    target = record.env[
+                        record["res_model" if "res_model" in record._fields else "model"]
+                    ].browse(record["res_id"])
+                else:
+                    target = record[relation_field.field_name]
+                data[relation_field.rename or relation_field.field_name] = (
+                    relation_field.replace_field_name_by_records(target)
+                )
+            self.add(record, data)
+
+    class Relation:
+        """Flags a record or field name to be added to the store in a relation."""
+
+        def __init__(
+            self,
+            records_or_field_names,
+            /,
+            as_thread=False,
+            only_id=False,
+            only_data=False,
+            rename=None,
```

Comment:
Note: `rename` is meant as a short term way to improve code without having to actually do the full rename in the client code in the same PR.